From: Rahul Rameshbabu <rrameshbabu@nvidia.com>
Date: Thu, 11 Sep 2025 16:02:19 -0700
Subject: backends/monitor-manager: Use normal modes when tiles are missing
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Originally, Meta.MonitorManager would handle tiled monitors without
gracefully handling the case where the main tile in the topology is
missing. This is problematic since, when setting the tiled topology, a
modeset will not be performed without the presence of the main tile. The
impact of this to an end user is having a black unusable screen on the
monitor. The issue is further emphasized by the fact that any detection of
tiled monitor support will try defaulting to tiling modes instead of normal
modes in the Meta.MonitorManager design, meaning the starting experience
using Mutter with such a monitor setup is a black unusable screen.

The situation where the main tile can be missing is when a monitor has
multiple physical connections. One connection has the main tile
characteristics in its EDID while the others have locations that
characterize them as not the main tile in the topology. If a user connects
to a port on the monitor that is not the main tile, no modeset is performed
and a black unusable screen is seen on the display.

Scenario 1: User plugs in all related connections to the monitor with
            tiling capabilities to the same GPU.

                    ┌────────────────────────────────┐
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │        Perfectly tiled!        │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    └─────────────┬─────┬────────────┘
             Port 1               │     │              Port 2
           main tile  ┌──────┐    │     │    ┌──────┐
                      ├──────┤    │     │    ├──────┤
                      │  DP  │ ───┴─────┴─── │  DP  │
                      │      │               │      │
                      └─┬──┬─┘               └─┬──┬─┘
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │
                        │  │                   │  │

Scenario 2: User plugs into the port of the monitor that is the main tile.

                    ┌────────────────────────────────┐
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │        Stretched content       │
                    │      due to missing tiles      │
                    │    or monitor fails to light   │
                    │     up based on its design     │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    └─────────────┬─────┬────────────┘
             Port 1               │     │              Port 2
           main tile  ┌──────┐    │     │
                      ├──────┤    │     │
                      │  DP  │ ───┴─────┴───
                      │      │
                      └─┬──┬─┘
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │

Scenario 3: User plugs into the port of the monitor that is not the main
            tile.

                    ┌────────────────────────────────┐
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │            Oh noes!            │
                    │          (no modeset)          │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    └─────────────┬─────┬────────────┘
             Port 1               │     │              Port 2
           main tile              │     │    ┌──────┐
                                  │     │    ├──────┤
                               ───┴─────┴─── │  DP  │
                                             │      │
                                             └─┬──┬─┘
                                               │  │
                                               │  │
                                               │  │
                                               │  │
                                               │  │
                                               │  │
                                               │  │
                                               │  │
                                               │  │

Scenario 4: User plugs into the port of the monitor that is the main tile.

                    ┌────────────────────────────────┐
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │        Stretched content       │
                    │      due to missing tiles      │
                    │    or monitor fails to light   │
                    │     up based on its design     │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    └─────────────┬─────┬────────────┘
             Port 1               │     │              Port 2
           main tile  ┌──────┐    │     │
                      ├──────┤    │     │
                      │  DP  │ ───┴─────┴───
                      │      │
                      └─┬──┬─┘
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │
                        │  │

With this change, scenarios 3 and 4, illustrated above, changes to the
following while scenario 1 and 2 remain the same.

                    ┌────────────────────────────────┐
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │       Content displayed        │
                    │            normally            │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    │                                │
                    └─────────────┬─────┬────────────┘
             Port 1               │     │              Port 2
           main tile              │     │    ┌──────┐
                                  │     │    ├──────┤
                               ───┴─────┴─── │  DP  │
                                             │      │
                                             └─┬──┬─┘
                                               │  │
                                               │  │
                                               │  │

Author: Rahul Rameshbabu <rrameshbabu@nvidia.com>
Bug-Ubuntu: https://bugs.launchpad.net/bugs/2131575
Origin: upstream, https://gitlab.gnome.org/GNOME/mutter/-/commit/b738453be835840c7bba01006f8ccd1975996a39
Forwarded: yes
Reviewed-By: Mitchell Augustin <mitchell.augustin@canonical.com>
Last-Update: 2025-11-19

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4685>
(cherry picked from commit b738453be835840c7bba01006f8ccd1975996a39)
Signed-off-by: Mitchell Augustin <mitchell.augustin@canonical.com>
---
 src/backends/meta-monitor-manager.c | 35 +++++++++++------
 src/backends/meta-monitor-private.h |  5 ++-
 src/backends/meta-monitor.c         | 75 ++++++++++++++++++++++++++++++++++++-
 3 files changed, 100 insertions(+), 15 deletions(-)

diff --git a/src/backends/meta-monitor-manager.c b/src/backends/meta-monitor-manager.c
index 7c4622c..531470b 100644
--- a/src/backends/meta-monitor-manager.c
+++ b/src/backends/meta-monitor-manager.c
@@ -4030,20 +4030,33 @@ rebuild_monitors (MetaMonitorManager *manager)
               if (is_main_tiled_monitor_output (output))
                 {
                   MetaMonitorTiled *monitor_tiled;
-
-                  monitor_tiled = meta_monitor_tiled_new (manager, output);
-                  manager->monitors = g_list_append (manager->monitors,
-                                                     monitor_tiled);
+                  g_autoptr (GError) error = NULL;
+
+                  monitor_tiled = meta_monitor_tiled_new (manager, output, &error);
+                  if (monitor_tiled == NULL)
+                    {
+                      g_warning ("Failed to create tiled monitor output: %s",
+                                 error->message);
+                    }
+                  else
+                    {
+                      manager->monitors = g_list_append (manager->monitors,
+                                                         monitor_tiled);
+                    }
                 }
             }
-          else
-            {
-              MetaMonitorNormal *monitor_normal;
+        }
+      for (k = meta_gpu_get_outputs (gpu); k; k = k->next)
+        {
+          MetaOutput *output = META_OUTPUT (k->data);
+          MetaMonitorNormal *monitor_normal;
 
-              monitor_normal = meta_monitor_normal_new (manager, output);
-              manager->monitors = g_list_append (manager->monitors,
-                                                 monitor_normal);
-            }
+          if (meta_output_get_monitor (output))
+            continue;
+
+          monitor_normal = meta_monitor_normal_new (manager, output);
+          manager->monitors = g_list_append (manager->monitors,
+                                             monitor_normal);
         }
     }
 
diff --git a/src/backends/meta-monitor-private.h b/src/backends/meta-monitor-private.h
index ec127bc..fe97606 100644
--- a/src/backends/meta-monitor-private.h
+++ b/src/backends/meta-monitor-private.h
@@ -103,8 +103,9 @@ G_DECLARE_FINAL_TYPE (MetaMonitorTiled, meta_monitor_tiled,
                       META, MONITOR_TILED,
                       MetaMonitor)
 
-MetaMonitorTiled * meta_monitor_tiled_new (MetaMonitorManager *monitor_manager,
-                                           MetaOutput         *output);
+MetaMonitorTiled * meta_monitor_tiled_new (MetaMonitorManager  *monitor_manager,
+                                           MetaOutput          *output,
+                                           GError             **error);
 
 MetaMonitorNormal * meta_monitor_normal_new (MetaMonitorManager *monitor_manager,
                                              MetaOutput         *output);
diff --git a/src/backends/meta-monitor.c b/src/backends/meta-monitor.c
index adda5a6..d0114dc 100644
--- a/src/backends/meta-monitor.c
+++ b/src/backends/meta-monitor.c
@@ -1784,9 +1784,69 @@ reset_tiled_monitor (MetaMonitorTiled *monitor_tiled,
   meta_monitor_init_supported_color_modes (monitor);
 }
 
+static gboolean
+verify_tiles_filled (GList *tile_outputs)
+{
+  uint32_t group_id = 0;
+  uint32_t max_h_tiles;
+  uint32_t max_v_tiles;
+  uint32_t tile_w;
+  uint32_t tile_h;
+  g_autofree gboolean *tiles = NULL;
+  GList *l;
+
+  for (l = tile_outputs; l; l = l->next)
+    {
+      MetaOutput *output = META_OUTPUT (l->data);
+      const MetaOutputInfo *output_info =
+        meta_output_get_info (output);
+      const MetaTileInfo *tile_info = &output_info->tile_info;
+
+      if (!tile_info->group_id)
+        return FALSE;
+
+      if (tile_info->loc_h_tile >= tile_info->max_h_tiles ||
+          tile_info->loc_v_tile >= tile_info->max_v_tiles)
+        return FALSE;
+
+      if (!group_id)
+        {
+          group_id = tile_info->group_id;
+          max_h_tiles = tile_info->max_h_tiles;
+          max_v_tiles = tile_info->max_v_tiles;
+
+          if ((max_h_tiles * max_v_tiles) != g_list_length (tile_outputs))
+            return FALSE;
+
+          tile_w = tile_info->tile_w;
+          tile_h = tile_info->tile_h;
+          tiles = g_new0 (gboolean, max_h_tiles * max_v_tiles);
+          tiles[tile_info->loc_h_tile +
+                (tile_info->loc_v_tile * max_h_tiles)] = TRUE;
+          continue;
+        }
+
+      if (group_id != tile_info->group_id ||
+          max_h_tiles != tile_info->max_h_tiles ||
+          max_v_tiles != tile_info->max_v_tiles ||
+          tile_w != tile_info->tile_w ||
+          tile_h != tile_info->tile_h)
+        return FALSE;
+
+      if (tiles[tile_info->loc_h_tile + (tile_info->loc_v_tile * max_h_tiles)])
+        return FALSE;
+
+      tiles[tile_info->loc_h_tile +
+            (tile_info->loc_v_tile * max_h_tiles)] = TRUE;
+    }
+
+  return TRUE;
+}
+
 MetaMonitorTiled *
-meta_monitor_tiled_new (MetaMonitorManager *monitor_manager,
-                        MetaOutput         *output)
+meta_monitor_tiled_new (MetaMonitorManager  *monitor_manager,
+                        MetaOutput          *output,
+                        GError             **error)
 {
   const MetaOutputInfo *output_info = meta_output_get_info (output);
   g_autolist (MetaOutput) outputs = NULL;
@@ -1809,6 +1869,14 @@ meta_monitor_tiled_new (MetaMonitorManager *monitor_manager,
   outputs = find_tiled_monitor_outputs (meta_output_get_gpu (output),
                                         origin_output);
 
+  if (!verify_tiles_filled (outputs))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Invalid tile group %u",
+                   meta_output_get_info (origin_output)->tile_info.group_id);
+      return NULL;
+    }
+
   main_output = find_untiled_output (origin_output, outputs);
 
   reset_tiled_monitor (monitor_tiled, g_steal_pointer (&outputs),
@@ -1977,6 +2045,9 @@ meta_monitor_tiled_update_outputs (MetaMonitor *monitor)
                             main_output))
     return FALSE;
 
+  if (!verify_tiles_filled (outputs))
+    return FALSE;
+
   reset_tiled_monitor (monitor_tiled, g_steal_pointer (&outputs),
                        origin_output, main_output);
 
